File: code/packages/python/actor/src/actor/actor_system.py
"""ActorSystem -- the runtime that manages actors, message delivery, and channels.

=== What Is an ActorSystem? ===

The ActorSystem is the office building. It has:

- A directory (which actors exist and their addresses)
- A mail room (message routing)
- A building manager (supervision -- restart actors that crash)

Actors are tenants. They register with the building, get an address, and the
building delivers their mail. But the building manager does not read the mail.

=== Operations ===

The ActorSystem provides five categories of operations:

1. Actor lifecycle: create_actor, stop_actor, get_actor_status
2. Messaging: send (enqueue a message in an actor's mailbox)
3. Processing: process_next (run one message through one actor's behavior),
   run_until_idle (process all actors round-robin until no work remains),
   run_until_done (like run_until_idle but keeps going until fully quiet)
4. Channels: create_channel, get_channel
5. Inspection: dead_letters, actor_ids, mailbox_size

=== Processing Model ===

In V1, actors are processed sequentially in round-robin order. The system
picks an actor with a non-empty mailbox, processes one message, then moves
to the next actor. This is simpler to test and debug than true parallelism.

    Round-robin processing:

    Actors:   [A: 3 msgs] [B: 1 msg] [C: 2 msgs]

    Step 1: Process A's first message  -> [A: 2] [B: 1] [C: 2]
    Step 2: Process B's first message  -> [A: 2] [B: 0] [C: 2]
    Step 3: Process C's first message  -> [A: 2] [B: 0] [C: 1]
    Step 4: Process A's second message -> [A: 1] [B: 0] [C: 1]
    Step 5: Process C's second message -> [A: 1] [B: 0] [C: 0]
    Step 6: Process A's third message  -> [A: 0] [B: 0] [C: 0]
    Done: all mailboxes empty.

=== Dead Letters ===

When a message cannot be delivered (the target actor does not exist or is
stopped), the message goes to the dead_letters list. This is a debugging
aid -- it tells you "these messages were sent but never processed."

=== Error Handling ===

If an actor's behavior function throws an exception during processing:

1. The error is logged
2. The message that caused the error goes to dead_letters
3. The actor's state is NOT changed (no partial updates)
4. The actor returns to IDLE status (continues processing next messages)
5. The system does NOT crash -- one misbehaving actor does not bring down
   the entire system
"""

from __future__ import annotations

import logging
from typing import TYPE_CHECKING, Any

from actor.actor import IDLE, STOPPED, Actor, ActorResult
from actor.channel import Channel

if TYPE_CHECKING:
    from collections.abc import Callable

    from actor.message import Message

logger = logging.getLogger(__name__)


# ---------------------------------------------------------------------------
# Exceptions
# ---------------------------------------------------------------------------


class ActorNotFoundError(Exception):
    """Raised when an operation targets a non-existent actor."""


class DuplicateActorError(Exception):
    """Raised when creating an actor with an id that already exists."""


class DuplicateChannelError(Exception):
    """Raised when creating a channel with an id that already exists."""


class ChannelNotFoundError(Exception):
    """Raised when getting a channel that does not exist."""


# ---------------------------------------------------------------------------
# ActorSystem
# ---------------------------------------------------------------------------


class ActorSystem:
    """Runtime for managing actors, message delivery, and channels.

    The ActorSystem is the "world" that actors live in. It manages:

    - Actor registration and lifecycle (create, stop, status)
    - Message delivery (send to an actor's mailbox)
    - Processing (run behavior functions, one message at a time)
    - Channels (create and retrieve append-only message logs)
    - Dead letters (messages that could not be delivered)

    === Quick Start ===

        >>> from actor.message import Message
        >>> from actor.actor import ActorResult
        >>>
        >>> def echo(state, message):
        ...     reply = Message.text("echo", f"echo: {message.payload_text}")
        ...     return ActorResult(state, [(message.sender_id, reply)])
        ...
        >>> system = ActorSystem()
        >>> system.create_actor("echo", None, echo)
        'echo'
        >>> system.send("echo", Message.text("user", "hello"))
        >>> system.process_next("echo")
        True
    """

    def __init__(self) -> None:
        """Create a new ActorSystem with no actors, no channels, and no dead letters."""
        self._actors: dict[str, Actor] = {}
        self._channels: dict[str, Channel] = {}
        self._dead_letters: list[Message] = []
        self._clock: int = 0

    # ------------------------------------------------------------------
    # Actor lifecycle
    # ------------------------------------------------------------------

    def create_actor(
        self,
        actor_id: str,
        initial_state: Any,
        behavior: Callable[[Any, Message], ActorResult],
    ) -> str:
        """Create and register a new actor.

        Args:
            actor_id: Unique identifier for the actor. Must not conflict
                with any existing actor.
            initial_state: The actor's starting state. Can be any type.
            behavior: Function that processes messages. Signature:
                (state, message) -> ActorResult.

        Returns:
            The actor_id (for convenience in chaining).

        Raises:
            DuplicateActorError: If an actor with this id already exists.

        Example:
            >>> system = ActorSystem()
            >>> system.create_actor("counter", 0, counter_behavior)
            'counter'
        """
        if actor_id in self._actors:
            raise DuplicateActorError(
                f"Actor with id '{actor_id}' already exists."
            )
        actor = Actor(actor_id, state=initial_state, behavior=behavior)
        self._actors[actor_id] = actor
        return actor_id

    def stop_actor(self, actor_id: str) -> None:
        """Stop an actor permanently.

        Sets the actor's status to STOPPED and drains its mailbox to
        dead_letters. A stopped actor cannot receive or process any
        more messages.

        Args:
            actor_id: The id of the actor to stop.

        Raises:
            ActorNotFoundError: If no actor with this id exists.
        """
        actor = self._actors.get(actor_id)
        if actor is None:
            raise ActorNotFoundError(
                f"Actor with id '{actor_id}' not found."
            )
        actor.status = STOPPED
        # Drain remaining mailbox messages to dead_letters
        drained = actor.drain_mailbox()
        self._dead_letters.extend(drained)

    def get_actor_status(self, actor_id: str) -> str:
        """Get an actor's current status.

        Args:
            actor_id: The id of the actor.

        Returns:
            One of: "idle", "processing", "stopped".

        Raises:
            ActorNotFoundError: If no actor with this id exists.
        """
        actor = self._actors.get(actor_id)
        if actor is None:
            raise ActorNotFoundError(
                f"Actor with id '{actor_id}' not found."
            )
        return actor.status

    # ------------------------------------------------------------------
    # Messaging
    # ------------------------------------------------------------------

    def send(self, target_id: str, message: Message) -> None:
        """Deliver a message to an actor's mailbox.

        If the target actor does not exist or is stopped, the message
        goes to dead_letters instead. This is a deliberate design choice:
        the sender should not crash because a recipient is unavailable.

        Args:
            target_id: The id of the actor to receive the message.
            message: The message to deliver.

        Behavior:
            - Target exists and is IDLE/PROCESSING: message enqueued
            - Target exists but is STOPPED: message goes to dead_letters
            - Target does not exist: message goes to dead_letters
        """
        actor = self._actors.get(target_id)
        if actor is None or actor.status == STOPPED:
            self._dead_letters.append(message)
            return
        actor.enqueue(message)

    # ------------------------------------------------------------------
    # Processing
    # ------------------------------------------------------------------

    def process_next(self, actor_id: str) -> bool:
        """Process one message from an actor's mailbox.

        Dequeues the front message, calls the actor's behavior function,
        applies the result (state update, message sends, actor creation),
        and updates the actor's status.

        If the behavior function raises an exception:
        - The actor's state is NOT changed
        - The failed message goes to dead_letters
        - The actor returns to IDLE (continues processing future messages)
        - The exception is logged but does NOT propagate

        Args:
            actor_id: The id of the actor to process.

        Returns:
            True if a message was processed, False if the mailbox was empty.

        Raises:
            ActorNotFoundError: If no actor with this id exists.
        """
        actor = self._actors.get(actor_id)
        if actor is None or actor.status == STOPPED:
            raise ActorNotFoundError(
                f"Actor with id '{actor_id}' not found or is stopped."
            )

        message = actor.dequeue()
        if message is None:
            return False

        # Set status to PROCESSING while the behavior runs
        actor.status = "processing"

        try:
            result = actor.behavior(actor.state, message)
        except Exception:
            # Behavior threw an exception. Per the spec:
            # 1. State is unchanged
            # 2. Message goes to dead_letters
            # 3. Actor returns to IDLE
            logger.exception(
                "Actor '%s' behavior raised an exception while processing "
                "message '%s'. State unchanged, message moved to dead_letters.",
                actor_id,
                message.id,
            )
            self._dead_letters.append(message)
            actor.status = IDLE
            return True

        # Apply the result
        actor.state = result.new_state

        # Deliver outgoing messages
        for target_id, msg in result.messages_to_send:
            self.send(target_id, msg)

        # Create new actors
        for spec in result.actors_to_create:
            # If create_actor fails (duplicate id), log and continue
            try:
                self.create_actor(spec.actor_id, spec.initial_state, spec.behavior)
            except DuplicateActorError:
                logger.warning(
                    "Actor '%s' tried to create actor '%s' which already exists.",
                    actor_id,
                    spec.actor_id,
                )

        # Update status based on stop flag
        if result.stop:
            actor.status = STOPPED
            # Drain remaining mailbox to dead_letters
            drained = actor.drain_mailbox()
            self._dead_letters.extend(drained)
        else:
            actor.status = IDLE

        return True

    def run_until_idle(self) -> dict[str, int]:
        """Process all actors round-robin until no work remains.

        Repeatedly finds an actor with a non-empty mailbox and IDLE status,
        processes one message, then moves to the next actor. Stops when
        no actor has pending messages.

        Returns:
            Statistics dict with keys:
                messages_processed: Total messages processed in this run
                actors_created: Total new actors created

        Example:
            >>> system = ActorSystem()
            >>> system.create_actor("echo", None, echo_behavior)
            'echo'
            >>> system.send("echo", Message.text("user", "hi"))
            >>> stats = system.run_until_idle()
            >>> stats["messages_processed"]
            1
        """
        messages_processed = 0
        actors_before = len(self._actors)

        while True:
            # Find any IDLE actor with messages
            found = False
            for actor_id, actor in list(self._actors.items()):
                if actor.status == IDLE and actor.mailbox_size > 0:
                    self.process_next(actor_id)
                    messages_processed += 1
                    found = True
                    break  # Restart the scan for round-robin fairness

            if not found:
                break

        actors_created = len(self._actors) - actors_before
        return {
            "messages_processed": messages_processed,
            "actors_created": max(0, actors_created),
        }

    def run_until_done(self) -> dict[str, int]:
        """Process all actors until the system is completely quiet.

        Like run_until_idle(), but keeps running until no new messages
        are being generated. This handles cases where processing a
        message generates new messages that need further processing.

        Returns:
            Statistics dict with the same keys as run_until_idle().
        """
        total_processed = 0
        actors_before = len(self._actors)

        while True:
            stats = self.run_until_idle()
            total_processed += stats["messages_processed"]
            if stats["messages_processed"] == 0:
                break

        actors_created = len(self._actors) - actors_before
        return {
            "messages_processed": total_processed,
            "actors_created": max(0, actors_created),
        }

    # ------------------------------------------------------------------
    # Channels
    # ------------------------------------------------------------------

    def create_channel(self, channel_id: str, name: str) -> Channel:
        """Create and register a new channel.

        Args:
            channel_id: Unique identifier for the channel.
            name: Human-readable name for the channel.

        Returns:
            The newly created Channel.

        Raises:
            DuplicateChannelError: If a channel with this id already exists.
        """
        if channel_id in self._channels:
            raise DuplicateChannelError(
                f"Channel with id '{channel_id}' already exists."
            )
        channel = Channel(channel_id, name)
        self._channels[channel_id] = channel
        return channel

    def get_channel(self, channel_id: str) -> Channel:
        """Retrieve a channel by its id.

        Args:
            channel_id: The id of the channel.

        Returns:
            The Channel object.

        Raises:
            ChannelNotFoundError: If no channel with this id exists.
        """
        channel = self._channels.get(channel_id)
        if channel is None:
            raise ChannelNotFoundError(
                f"Channel with id '{channel_id}' not found."
            )
        return channel

    # ------------------------------------------------------------------
    # Inspection
    # ------------------------------------------------------------------

    @property
    def dead_letters(self) -> list[Message]:
        """Messages that could not be delivered.

        Returns a copy to prevent external modification.
        """
        return list(self._dead_letters)

    def actor_ids(self) -> list[str]:
        """List all registered actor IDs.

        Returns:
            A list of actor id strings.
        """
        return list(self._actors.keys())

    def mailbox_size(self, actor_id: str) -> int:
        """Get the number of pending messages for an actor.

        Args:
            actor_id: The id of the actor.

        Returns:
            The number of messages in the actor's mailbox.

        Raises:
            ActorNotFoundError: If no actor with this id exists.
        """
        actor = self._actors.get(actor_id)
        if actor is None:
            raise ActorNotFoundError(
                f"Actor with id '{actor_id}' not found."
            )
        return actor.mailbox_size

    # ------------------------------------------------------------------
    # Shutdown
    # ------------------------------------------------------------------

    def shutdown(self) -> None:
        """Stop all actors and drain all mailboxes to dead_letters.

        This is a clean shutdown: every actor is stopped, every unprocessed
        message is preserved in dead_letters for debugging.
        """
        for actor_id in list(self._actors.keys()):
            actor = self._actors[actor_id]
            if actor.status != STOPPED:
                actor.status = STOPPED
                drained = actor.drain_mailbox()
                self._dead_letters.extend(drained)

    def __repr__(self) -> str:
        """Human-readable representation showing actor and channel counts."""
        return (
            f"ActorSystem(actors={len(self._actors)}, "
            f"channels={len(self._channels)}, "
            f"dead_letters={len(self._dead_letters)})"
        )

File: code/packages/go/actor/actor_system.go
package actor

// ActorSystem is the runtime that manages actor lifecycles, message delivery,
// and channels. It is the "world" that actors live in.
//
// # Analogy
//
// The ActorSystem is the office building. It has a directory (which actors
// exist and their addresses), a mail room (message routing), and a building
// manager (supervision — restart actors that crash). Actors are tenants.
// They register with the building, get an address, and the building delivers
// their mail. But the building manager does not read the mail.
//
// # Processing Model
//
// In V1, the ActorSystem processes actors sequentially in round-robin order.
// True parallelism (multiple actors processing simultaneously on different
// goroutines) is a future enhancement. The sequential model is simpler to
// test, debug, and reason about.
//
// # Dead Letters
//
// When a message cannot be delivered — because the target actor doesn't exist
// or has been stopped — the message goes to the dead letter queue. This is
// useful for debugging: you can inspect dead letters to find routing errors,
// missing actors, or messages sent after shutdown.

import (
	"fmt"
)

// ─────────────────────────────────────────────────────────────────────────────
// Errors
// ─────────────────────────────────────────────────────────────────────────────

// ErrActorExists is returned when creating an actor with an ID that is
// already registered. Actor IDs must be unique within a system.
var ErrActorExists = fmt.Errorf("actor: actor with this ID already exists")

// ErrActorNotFound is returned when sending to or processing an actor
// that is not registered in the system.
var ErrActorNotFound = fmt.Errorf("actor: actor not found")

// ─────────────────────────────────────────────────────────────────────────────
// ActorSystem
// ─────────────────────────────────────────────────────────────────────────────

// ActorSystem manages actor lifecycles, message routing, and channels.
//
// Fields:
//
//	+─────────────+──────────────────────────────────────────────────+
//	| actors      | Map of actor_id → Actor. The registry.          |
//	| channels    | Map of channel_id → Channel.                    |
//	| deadLetters | Messages that could not be delivered.            |
//	| clock       | Monotonic counter for message timestamps.        |
//	| actorOrder  | Insertion-ordered list of actor IDs for          |
//	|             | deterministic round-robin processing.            |
//	+─────────────+──────────────────────────────────────────────────+
type ActorSystem struct {
	actors      map[string]*Actor
	channels    map[string]*Channel
	deadLetters []*Message
	clock       int64
	actorOrder  []string
}

// NewActorSystem creates a new empty actor system.
//
//	system := NewActorSystem()
//	system.CreateActor("echo", nil, echoBehavior)
//	system.Send("echo", NewTextMessage("user", "hello", nil))
//	system.RunUntilIdle()
func NewActorSystem() *ActorSystem {
	return &ActorSystem{
		actors:      make(map[string]*Actor),
		channels:    make(map[string]*Channel),
		deadLetters: make([]*Message, 0),
		clock:       0,
		actorOrder:  make([]string, 0),
	}
}

// ─────────────────────────────────────────────────────────────────────────────
// Actor Lifecycle
// ─────────────────────────────────────────────────────────────────────────────

// CreateActor registers a new actor in the system with the given ID,
// initial state, and behavior function.
//
// Returns the actor ID on success. Returns ErrActorExists if an actor
// with the same ID already exists.
//
// The actor starts in IDLE status with an empty mailbox.
//
//	id, err := system.CreateActor("counter", 0, counterBehavior)
func (s *ActorSystem) CreateActor(id string, initialState interface{}, behavior BehaviorFunc) (string, error) {
	if _, exists := s.actors[id]; exists {
		return "", fmt.Errorf("%w: %s", ErrActorExists, id)
	}
	a := newActor(id, initialState, behavior)
	s.actors[id] = a
	s.actorOrder = append(s.actorOrder, id)
	return id, nil
}

// StopActor sets the actor's status to STOPPED and drains any remaining
// messages in its mailbox to dead letters.
//
// After stopping, the actor will not accept new messages — any messages
// sent to it will go to dead letters.
//
// Returns ErrActorNotFound if no actor with the given ID exists.
func (s *ActorSystem) StopActor(id string) error {
	a, exists := s.actors[id]
	if !exists {
		return fmt.Errorf("%w: %s", ErrActorNotFound, id)
	}

	a.status = StatusStopped

	// Drain remaining mailbox messages to dead letters.
	// These messages will never be processed — save them for debugging.
	for {
		msg := a.dequeue()
		if msg == nil {
			break
		}
		s.deadLetters = append(s.deadLetters, msg)
	}

	return nil
}

// GetActorStatus returns the lifecycle status of the actor with the given ID.
// Returns StatusIdle, StatusProcessing, or StatusStopped.
//
// Returns an error if the actor is not found.
func (s *ActorSystem) GetActorStatus(id string) (ActorStatus, error) {
	a, exists := s.actors[id]
	if !exists {
		return "", fmt.Errorf("%w: %s", ErrActorNotFound, id)
	}
	return a.status, nil
}

// ─────────────────────────────────────────────────────────────────────────────
// Messaging
// ─────────────────────────────────────────────────────────────────────────────

// Send delivers a message to the target actor's mailbox.
//
// If the target actor does not exist or is stopped, the message goes to
// dead letters and an error is returned.
//
// The message delivery algorithm:
//  1. Look up target_id in actors map.
//  2. If NOT FOUND → dead_letters.append(message). Return error.
//  3. If STOPPED → dead_letters.append(message). Return error.
//  4. Enqueue message in target's mailbox.
//
// Time complexity: O(1) — hash map lookup + queue append.
func (s *ActorSystem) Send(targetID string, msg *Message) error {
	a, exists := s.actors[targetID]
	if !exists {
		s.deadLetters = append(s.deadLetters, msg)
		return fmt.Errorf("%w: %s", ErrActorNotFound, targetID)
	}
	if a.status == StatusStopped {
		s.deadLetters = append(s.deadLetters, msg)
		return fmt.Errorf("%w: %s (stopped)", ErrActorNotFound, targetID)
	}
	a.enqueue(msg)
	return nil
}

// ─────────────────────────────────────────────────────────────────────────────
// Processing
// ─────────────────────────────────────────────────────────────────────────────

// ProcessNext processes one message from the specified actor's mailbox.
//
// Returns true if a message was processed, false if the mailbox was empty.
// Returns an error if the actor is not found or is stopped.
//
// The processing algorithm:
//  1. Look up actor. Return error if not found or stopped.
//  2. If mailbox empty, return false (no work).
//  3. Dequeue front message.
//  4. Set status = PROCESSING.
//  5. Call behavior(state, message) → ActorResult.
//  6. If behavior returns error: state unchanged, message → dead letters,
//     actor continues processing (status = IDLE).
//  7. Update state to ActorResult.NewState.
//  8. Deliver each outgoing message via Send().
//  9. Create each new actor from ActorsToCreate specs.
//  10. If ActorResult.Stop: set status = STOPPED, drain mailbox.
//  11. Else: set status = IDLE.
func (s *ActorSystem) ProcessNext(actorID string) (bool, error) {
	a, exists := s.actors[actorID]
	if !exists {
		return false, fmt.Errorf("%w: %s", ErrActorNotFound, actorID)
	}
	if a.status == StatusStopped {
		return false, fmt.Errorf("%w: %s (stopped)", ErrActorNotFound, actorID)
	}

	// If mailbox is empty, there's nothing to process.
	msg := a.dequeue()
	if msg == nil {
		return false, nil
	}

	// Set status to PROCESSING while the behavior function runs.
	a.status = StatusProcessing

	// Call the behavior function.
	result, err := a.behavior(a.state, msg)
	if err != nil {
		// Behavior threw an error:
		//   - State is UNCHANGED (we don't apply partial updates)
		//   - Message goes to dead letters for debugging
		//   - Actor continues processing (returns to IDLE)
		s.deadLetters = append(s.deadLetters, msg)
		a.status = StatusIdle
		return true, nil
	}

	// Apply the result.
	a.state = result.NewState

	// Deliver outgoing messages.
	for _, out := range result.MessagesToSend {
		// Ignore send errors — they'll land in dead letters automatically.
		_ = s.Send(out.TargetID, out.Msg)
	}

	// Create new actors.
	for _, spec := range result.ActorsToCreate {
		_, _ = s.CreateActor(spec.ID, spec.InitialState, spec.Behavior)
	}

	// Check if the actor wants to stop.
	if result.Stop {
		a.status = StatusStopped
		// Drain remaining mailbox to dead letters.
		for {
			remaining := a.dequeue()
			if remaining == nil {
				break
			}
			s.deadLetters = append(s.deadLetters, remaining)
		}
	} else {
		a.status = StatusIdle
	}

	return true, nil
}

// RunUntilIdle processes all actors in round-robin order until no actor has
// pending messages.
//
// The algorithm:
//  1. Scan all actors for one with status==IDLE and non-empty mailbox.
//  2. If none found, return (system is idle).
//  3. Process one message from that actor.
//  4. Repeat from step 1.
//
// Returns statistics about what happened during processing.
//
// NOTE: In V1, this processes actors one at a time. True parallelism
// is a future enhancement.
func (s *ActorSystem) RunUntilIdle() map[string]int {
	stats := map[string]int{
		"messages_processed": 0,
		"actors_created":     0,
	}

	for {
		found := false
		for _, id := range s.actorOrder {
			a, exists := s.actors[id]
			if !exists {
				continue
			}
			if a.status == StatusIdle && len(a.mailbox) > 0 {
				actorsBefore := len(s.actors)
				processed, _ := s.ProcessNext(id)
				if processed {
					stats["messages_processed"]++
					actorsAfter := len(s.actors)
					stats["actors_created"] += actorsAfter - actorsBefore
					found = true
				}
			}
		}
		// Also check any newly created actors (their IDs were appended
		// to actorOrder during this round).
		if !found {
			break
		}
	}

	return stats
}

// RunUntilDone processes all actors repeatedly until the system is completely
// quiet — no messages in any mailbox and no new messages being generated.
//
// This is like RunUntilIdle but it keeps going through multiple rounds.
// RunUntilIdle stops when one pass finds no work. RunUntilDone repeats
// until the system truly settles.
//
// Safety: This will loop forever if actors keep generating new messages
// indefinitely (e.g., ping-pong without a stop condition). Use with caution.
func (s *ActorSystem) RunUntilDone() map[string]int {
	totalStats := map[string]int{
		"messages_processed": 0,
		"actors_created":     0,
	}

	for {
		stats := s.RunUntilIdle()
		totalStats["messages_processed"] += stats["messages_processed"]
		totalStats["actors_created"] += stats["actors_created"]

		if stats["messages_processed"] == 0 {
			break
		}
	}

	return totalStats
}

// ─────────────────────────────────────────────────────────────────────────────
// Channels
// ─────────────────────────────────────────────────────────────────────────────

// CreateChannel creates and registers a new channel in the actor system.
//
//	ch := system.CreateChannel("ch_001", "email-summaries")
func (s *ActorSystem) CreateChannel(id, name string) *Channel {
	ch := NewChannel(id, name)
	s.channels[id] = ch
	return ch
}

// GetChannel retrieves a channel by ID. Returns nil if not found.
func (s *ActorSystem) GetChannel(id string) *Channel {
	return s.channels[id]
}

// ─────────────────────────────────────────────────────────────────────────────
// Inspection
// ─────────────────────────────────────────────────────────────────────────────

// DeadLetters returns a copy of all messages that could not be delivered.
// These include messages sent to non-existent actors, stopped actors,
// and messages that caused behavior function errors.
func (s *ActorSystem) DeadLetters() []*Message {
	result := make([]*Message, len(s.deadLetters))
	copy(result, s.deadLetters)
	return result
}

// ActorIDs returns a list of all registered actor IDs.
func (s *ActorSystem) ActorIDs() []string {
	ids := make([]string, 0, len(s.actors))
	for id := range s.actors {
		ids = append(ids, id)
	}
	return ids
}

// MailboxSize returns the number of pending messages in the specified
// actor's mailbox. Returns 0 if the actor is not found.
func (s *ActorSystem) MailboxSize(actorID string) int {
	a, exists := s.actors[actorID]
	if !exists {
		return 0
	}
	return len(a.mailbox)
}

// Shutdown stops all actors, drains all mailboxes to dead letters.
func (s *ActorSystem) Shutdown() {
	for _, id := range s.actorOrder {
		a, exists := s.actors[id]
		if !exists {
			continue
		}
		if a.status != StatusStopped {
			_ = s.StopActor(id)
		}
	}
}
